### PR TITLE
OCPBUGS-1959: show alert on dev catalog page when chart repo is unreachable/incorrect

### PR DESCRIPTION
**Pull request by debsmita1**
**Fixes:**
https://issues.redhat.com/browse/OCPBUGS-1959

**Root analysis:**
- on the dev catalog page, the helm charts in other chart repositories are hidden if a chart repository is unreachable/incorrect
- on the helm catalog page, `Error loading catalog items` is shown if a chart repository is unreachable/incorrect

**Solution Description:**
- show charts from other chart repositories if a chart repository is unreachable/incorrect
- show an inline alert on the developer catalog page to inform the user about the chart that is unreachable/incorrect
- show a warning icon next to the chart repositories filter on the helm catalog page when there is an invalid chart repo
- show a warning icon next to the sub-catalog type when an entire sub-catalog in unreachable

**Test setup:**
-> Goto Helm page
-> Click on `Create` dropdown on the right
-> Select `Repository`
-> Create Helm Chart Repository with chart url "https://github.com/test"

**Screenshot/GIF:**
1. When there is an invalid chart repo, and the builder images are not loaded
Screenshot 2022-11-09 at 6 51 59 PM

2. On the helm catalog page, when there is an invalid chart repo
Screenshot 2022-11-09 at 5 17 55 PM

/kind bug

### CODE REVIEW

**Review by jerolimov — CHANGES_REQUESTED** [review]

**Comment by serenamarie125** [comment]
@debsmita1 thanks for addressing this!  couple of suggestions/questions:

1. I think this should be a warning message, since there is missing information. WDYT?
2. I suggest changing the inline alert text to be **"The following Helm Chart Repositories are unreachable or invalid: xxx"**
3. in the case where test-repo was invalid/unreachable, i'm wondering if we should show the repository under Chart repositories filter, but be disabled including (0) and provide a warning symbol with a tooltip saying **"This Helm Chart Repository is unreachable or invalid."**
4. For Builder Images, do we know the difference when no builder images are found versus there are loading failures?
5. If so ... "No Builder Images are available at this time due to loading failures."
6. In the case where an entire subcatalog in unreachable, could we keep it under Type, have it disabled including (0) & provide warning symbol with a tooltip saying **"No Builder Images are available at this time due to loading failures."**

Happy to have a quick slack or call to discuss if this doesn't make sense or if you disagree :)

**Review by serenamarie125 — CHANGES_REQUESTED** [review]
Added some suggestions/requests.

**Comment by debsmita1** [comment]
@serenamarie125 

> 1. I think both of these should be a ⚠️ warning rather than ℹ️ information, since there is missing information. WDYT?

Agreed !

> 3. in the case where test-repo was invalid/unreachable, i'm wondering if we should show the repository under Chart repositories filter, but be disabled including (0) and provide a ⚠️ warning symbol with a tooltip saying **"This Helm Chart Repository is unreachable or invalid."**
 
 As of today, we evaluate the chart repositories filter based on the catalog items that are shown in the catalog

> 4. For Builder Images, do we know the difference when no builder images are found versus there are loading failures?

If no builder images are found, then we simply don't show anything related to builder images. If there is any error we show `Builder Images content is not available in the catalog at this time due to loading failures.` as an inline alert

> 6. In the case where an entire subcatalog in unreachable, could we keep it under Type, have it disabled including (0) & provide ⚠️ warning symbol with a tooltip saying **"No Builder Images are available at this time due to loading failures."**

Are you suggesting that instead of showing the inline alert at the top, we show a warning icon next to the sub-catalog type ?

**Comment by serenamarie125** [comment]
@debsmita1 
> As of today, we evaluate the chart repositories filter based on the catalog items that are shown in the catalog

What I was suggesting is ALSO showing a warning icon near the helmchart repo filter item  

> If no builder images are found, then we simply don't show anything related to builder images. If there is any error we show Builder Images content is not available in the catalog at this time due to loading failures. as an inline alert

Ah I see.  Can we change the message to "⚠️ No Builder Images are available at this time due to loading failures."

> Are you suggesting that instead of showing the inline alert at the top, we show a warning icon next to the sub-catalog type ?
I'm suggesting maybe we do it in addition

Can you update the GIF showing the changes from ℹ️ to ⚠️  as well?

Let me know if you have questions, I have time to meet tomorrow if we need to sync on a call.

**Comment by debsmita1** [comment]
> > As of today, we evaluate the chart repositories filter based on the catalog items that are shown in the catalog
> 
> What I was suggesting is ALSO showing a warning icon near the helmchart repo filter item
> 
@serenamarie125  We will not be able to show the warning icon next to the repo filter item because the chart repositories filter is shown based on the charts that are displayed in the catalog. So I could come up with this:
Screenshot 2022-11-09 at 5 17 55 PM
 
 Let me know what you think about this. I have updated the screens in the PR description

**Comment by debsmita1** [comment]
> What happens in the Dev Catalog when there are issues with Helm Chart Repos? I think we need to expose the warning icon there as well. WDYT @debsmita1

Showing the warning icon next to Helm Charts
Screenshot 2022-11-15 at 1 30 14 PM

**Review by serenamarie125 — APPROVED** [review]
Looks great @debsmita1 thanks for working through all these details !

File: frontend/packages/console-app/src/components/flags/FeatureFlagExtensionHookResolver.tsx [hunk 1]
```diff
@@ -10,7 +10,7 @@ const FeatureFlagExtensionHookResolver: React.FC<FeatureFlagExtensionHookResolve
   handler,
   setFeatureFlag,
 }) => {
-  handler(setFeatureFlag);
+  handler && handler(setFeatureFlag);
```
Comment (jerolimov) [inline frontend/packages/console-app/src/components/flags/FeatureFlagExtensionHookResolver.tsx:13]: Please mark the handler as optional if it is optional. Btw you can also run such a call with `handler?.(setFeatureFlag)`, but I'm also fine with this condition here.

File: frontend/packages/console-shared/src/components/catalog/CatalogController.tsx [hunk 2]
```diff
@@ -127,16 +128,22 @@ const CatalogController: React.FC<CatalogControllerProps> = ({
   }, [items, queryParams]);
 
   const catalogTypes: CatalogType[] = React.useMemo(() => {
-    const types = catalogExtensions
+    const types: CatalogType[] = catalogExtensions
       .map((extension) => ({
```
Comment (jerolimov) [inline frontend/packages/console-shared/src/components/catalog/CatalogController.tsx:132]: :tada: yeah types!

It's fine here, just for your info, you can alternative add the "expected return type" of the map as argument to map like this:

```suggestion
    const types = catalogExtensions
      .map<CatalogType>((extension) => ({
```

What's the difference? You can see this when changing label to label2 for example in the inner function.

When you say `types` is of type `CatalogType[]` you'll get an error on the assignment to `types`.

When you say `map` should return a type `CatalogType` you'll get an error on the arrow function that returns the wrong data.

Just FYI. It's fine to keep the existing code!

File: frontend/packages/console-shared/src/components/catalog/catalog-view/CatalogTypeSelector.tsx [hunk 3]
```diff
@@ -51,7 +54,32 @@ const CatalogTypeSelector: React.FC<CatalogTypeSelectorProps> = ({
...
+                    content={
+                      type.catalogTypeError instanceof Error ? type.catalogTypeError.message : null
+                    }
+                  >
+                    <YellowExclamationTriangleIcon />
+                  </Tooltip>
+                )}
+              </Link>
+            </li>
+          ) : typeCount === 0 && type.catalogTypeError ? (
```
Comment (jerolimov) [inline frontend/packages/console-shared/src/components/catalog/catalog-view/CatalogTypeSelector.tsx:70]: Both conditions are almost the same or?

Couldn't we build it like this:

```tsx
if (typeCount === 0 && !type.catalogTypeError) return null;

return (
  <li>
    <Link to={to}>{`${label} (${typeCount})`}</Link>
    {type.catalogTypeError && (
      <>
        {' '}
        <Tooltip content={
          typeCount === 0 ?
            t() :
            type.catalogTypeError?.message
        }>
          <YellowExclamationTriangleIcon />
        </Tooltip>
      </>
    )}
  </li>
)
```

Comment (debsmita1) [reply frontend/packages/console-shared/src/components/catalog/catalog-view/CatalogTypeSelector.tsx]: the labels with 0 count should not be a link

**Review by jerolimov — CHANGES_REQUESTED** [review]
Sorry for the long delay here. I looked into this PR several times but it is a tricky issue!

I added some code recommendations and I think we should take a look together, as I'm also not 100% sure. But I want share my findings here already.

## A: Enabled OpenShift helm charts + one broken helm repo

When I hope All Service the warning is shown and 46 helm charts. But when I click on Helm Charts (from the All Service) there is no warning anymore.

When I navigate back to the add page and click on Helm charts directly I get a warning.

Also when I reload on the helm subcatalog the warning is shown. Its just not there when I navigate from All Services to Helm.

https://user-images.githubusercontent.com/139310/210461184-73af2ee2-d5e9-4a5a-9696-dabae758f0c4.mp4

## B: Disabled OpenShift helm charts + one broken helm repo (0 charts)

When I disable also the default helm charts I see 0 helm charts and get a warning.

But when I click on the add page and then on Helm Charts I don't get any warning. Its just an empty page. Can we show the alert there as well?

https://user-images.githubusercontent.com/139310/210461295-28ee5e4c-9603-4233-886c-58c2a65856f0.mp4

I'm **_not sure_** if it would help to throw an `IncompleteError` directly when found an `conole-warning` ?!

**Pull request closed without merging**